### DISCUSSION TITLE
The dark side of model-viewer?

### DISCUSSION THREAD

**Discussion by Lemovision** [google/model-viewer#2869] [General]
I've been playing around with other viewers and looking back it seems model-viewer is really DARK to me
Looking at your [fidelity test page](https://modelviewer.dev/fidelity/), I think also in pretty much every single case it's the darkest of all
But I think more than dark shadows, I think the contrast between highlights and shadows might be too much
So the blacks seem just crushed in model-viewer and just can't show much detail in those areas

For example here both the shadows (black rubber) are under-exposed, and the highlights (wood top) overexposed
https://modelviewer.dev/fidelity/results/khronos-FlightHelmet/model-viewer-golden.png
https://modelviewer.dev/fidelity/results/khronos-FlightHelmet/babylon-golden.png

Also I was trying to have some realistic metal for my model but I think the highlight rolloff is just too abrupt like here
https://modelviewer.dev/fidelity/results/khronos-SpecGlossVsMetalRough/model-viewer-golden.png
https://modelviewer.dev/fidelity/results/khronos-SpecGlossVsMetalRough/babylon-golden.png

So it would be nice to have a less dark/contrasted look by default I think like the other viewers
And eventually have some post processing option to adjust the final contrast if needed

**Comment by elalish** [comment]
Thanks for the feedback! First, I'd like to point you to our `exposure` [attribute](https://modelviewer.dev/docs/index.html#entrydocs-lightingandenv-attributes-exposure), which gives some control. As for the fidelity results, try not to over-index on particular camera views. Those highlights  are very sensitive to the exact orientation of the face relative to the light. If you look  at them interactively, you'll  find they seem much more  similar than that  one shot implies (part of  the choice of fidelity shot is to highlight differences so we can reduce them). 

Still, yes we are looking into the reason  behind the *slightly* darkened renders in MV. It likely has to  do with details  of the  tone mapping which haven't been standardized in glTF yet. In general, just like with photography, the most  important part of getting the look you want is choosing the  proper  lighting. It's unlikely our one-size-fits-all default lighting is perfect for everything, so you  may  want to look into building custom HDR environments.

**Comment by Lemovision** [comment]
Yes I played with the exposure setting also in your model editor
But since the problem I think comes from the contrast, or tone mapping as you said
Even if I boost the exposure it stills results as a brighter over-contrasted image
(Btw I maxed it out at 2 in your editor, is the hard limit 2 or higher?)
For my own example here's the same model and same material same hdr file
It's a 3d scan I just made the other day in a nice temple here in Thailand
The one one the right looks a lot more realistic to me with less contrast
(LEFT https://modelviewer.dev/editor/ and RIGHT https://sandbox.babylonjs.com/)

![viewercompare](https://user-images.githubusercontent.com/61175245/137221415-59e4c030-c992-480c-8c08-d8b9327c0380.jpg)

**Reply by elalish** [reply]
Thanks for the comparison! I'm not sure the  difference is entirely contrast, though. It looks like  the surface  is  rougher in babylon; what  roughness value (or range) does this  use? That's one of the trickiest parts of rendering convergence. 

Also, to compare apples to apples, have  you set the babylon  sandbox to use ACES tone mapping?

**Reply by Lemovision** [reply]
They're both at metallic=1, roughness=0.1 for this one

Okay so the tone mapping in babylon was off
If I put ACES it's indeed more contrasted, although still a bit smoother I think
Can we have an option to disable tone mapping then in model-viewer too?

Babylon ACES
![viewercompareaces](https://user-images.githubusercontent.com/61175245/137223579-c34d5102-9f73-4820-bec8-1e08fdf195f1.jpg)

**Reply by elalish** [reply]
Disabling tone mapping entirely tends to give really poor results; try again with a full-sun HDR. The roughness difference is interesting; there's no  such thing as "looks better" in PBR, only "is more accurate". It would be really helpful to get a high quality path-traced  rendering  of this to check who  is closer. 

Babylon does a slow pre-process on their environments while we do that step real-time, so there may also be some tradeoff in performance vs accuracy. 

I'll add tone mapping  controls when  I find an easy way to explain it to lay  users! We're  definitely thinking about it though.

**Reply by Lemovision** [reply]
Okay thank you for having a look
I will do more testing as well
Cheers

**Comment by Lemovision** [comment]
Btw also same contrast difference with non-metallic rough material

![viewercomparerough](https://user-images.githubusercontent.com/61175245/137222801-fba6186d-0dc4-4862-9b43-1be33bd83f6c.jpg)

**Reply by elalish** [reply]
ah, that's helpful; again are you  using  consistent tone mapping?

**Reply by Lemovision** [reply]
Okay here's the same with ACES in babylon
But yeah would be nice to have tone mapping controls I think

![viewercompareroughaces](https://user-images.githubusercontent.com/61175245/137224817-ca65a597-3d2f-408f-b7a8-08805edc8e94.jpg)

**Comment by Lemovision** [comment]
Hi have another "dark side" question slightly related to this
There's a rendering thing that I noticed as well on reflective materials
On the edges it would reflect the dark areas of the environment map
I find this outline effect especially problematic on bright backgrounds
You would expect to reflect white on the edges or whatever background color
This is with your example model without textures in the model editor
I pointed out the most obvious parts, but I think it's all around more or less

![edges1](https://user-images.githubusercontent.com/61175245/138338641-9d35efc2-75e3-4c36-8dc6-32816225d24b.jpg)
![edges2](https://user-images.githubusercontent.com/61175245/138338647-4ef079d0-bee9-44ea-8ec2-46fd1ee823d8.jpg)

Would you have any way not to reflect the dark areas as much?
Or maybe have the background color affect the env/reflections a bit somehow
I think it would fit better in the background without these dark outlines

**Reply by elalish** [reply]
Thanks, this looks like it could be an error with the Fresnel term. @WestLangley any thoughts, since you've been looking at these shaders recently?

**Comment by Lemovision** [comment]
![ele1](https://user-images.githubusercontent.com/61175245/138360203-173a322a-bdbc-4845-9b59-e9bee84e5497.jpg)
![ele2](https://user-images.githubusercontent.com/61175245/138360205-c138facf-d9d7-461a-859f-752c024dfcbc.jpg)
![ele3](https://user-images.githubusercontent.com/61175245/138360204-69410e79-2478-40bd-aee5-d81433dd84a3.jpg)

No texture I swear 😅 just white material and neutral env map

**Comment by lanxinger** [comment]
We also find this to be the case. 

Take a look at how the same model is rendered in SceneKit with our App and how the same model is rendered in ModelViewer.

SceneKit:
![IMG_2881](https://user-images.githubusercontent.com/20549552/220964932-cd1b409c-d1e0-47c3-98c9-d1c3af185c33.PNG)

ModelViewer:
![IMG_2882](https://user-images.githubusercontent.com/20549552/220965089-e1c789a2-6fc8-45c9-a137-0b91261b1f2a.PNG)

Is there any adjustments we can make to make it less contrasty. I feel that just having exposure control alone is not enough.

**Reply by elalish** [reply]
To get consistent rendering, you'll need to apply the same environment-image to MV as SceneKit uses. Set the HDR you desire with `environment-image`.

**Comment by lanxinger** [comment]
thank you @elalish, understood. that means that we will most likely have to add the same HDR/EXR on both iPhone and web which will drastically increase our app size and loading times on web. I checked and at 4K they are about 23MB. What resolution HDRI does model viewer use as the default? is 1K or 2K sufficient these days?

**Reply by elalish** [reply]
We generally recommend 1K, though it actually depends strongly on your model. The price of a lower-res HDR is that it increases the minimum roughness. For the example you show, you could probably use an extremely low-res version (maybe even 265x128?) since it's fairly rough all over. You can try different ones in our editor to see at what point you actually notice a difference. Generally the models that really need high-res HDRs are electronics that have flat, shiny screens, especially if they are rendered at high resolution.

**Comment by lanxinger** [comment]
Thank you. Appreciate the advice. One last question. I checked but couldn’t find it in the docs. Is it possible to rotate the HDRI? I couldn’t find any dedicated option that you provide. It seems that using the same HDRi in SceneKit and model viewer is using a different rotation as a default so I want to match it.

**Reply by elalish** [reply]
Not trivially, but it's possible using `turntableRotation`, see this example: https://modelviewer.dev/examples/stagingandcameras/#turnSkybox
